File: codes/a_further_implement_by_java/CS304_SE_Project/src/main/java/Controllers/BackendController.java
package Controllers;

import States.DebuggingState;
import States.FileProcessingState;
import States.State;
import States.WaitingState;

import java.util.ArrayList;
import java.util.HashMap;

public class BackendController {
    private static final BackendController instance = new BackendController();

    State cur_state;
    HashMap<String, State> stateMap = new HashMap<>(){{
        put("WaitingState", new WaitingState());
        put("FileProcessingState", new FileProcessingState());
        put("DebuggingState", new DebuggingState());
    }};


    private BackendController() {
        cur_state = null;
    }

    public static BackendController getInstance() {
        return instance;
    }

    public void setState(String state) {
        cur_state = stateMap.get(state);
    }

    public void setState(State state) {
        cur_state = state;
    }

    public void doStateAction() {
        if (cur_state == null) {
            cur_state = new WaitingState();
        }
        cur_state.doAction();
    }
}

File: codes/CPU_Debugger/Controllers/BackendController.cpp
#include "BackendController.h"
#include "States/State.h"
#include "States/WaitingState.h"
#include "States/DebugginState.h"
#include <iostream>

BackendController::BackendController()
{
    this->stateMap.insert("Waiting", std::make_shared<WaitingState>());
    this->stateMap.insert("Debugging", std::make_shared<DebugginState>());
    curState = this->stateMap.value("Waiting");
}

BackendController& BackendController::getInstance()
{
    static BackendController instance;
    return instance;
}

void BackendController::setState(QString string){
    if (!stateMap.contains(string))
    {
        std::cout << "Please check the state name is right." << std::endl;
        return;
    }
    this->curState = this->stateMap.value(string);
}

std::shared_ptr<State> BackendController::getState()
{
    return this->curState;
}

void BackendController::doStateAction()
{
    this->curState->doAction();
}
